keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -0,0 +1,1190 @@
+-->
+
+At a high level, CR validation in kubernetes has three components:
+1. `kube-openapi`-based `Validate` using `*spec.Schema`
+2. StructuralSchema-based Meta, TypeMeta, ListSet/Map validation
+3. CEL Validators using a CEL schema interface type
+
+Each of these validation paths will need to have their `ValidateUpdate(old, new)`
+codepath modified to:
+
+1. Take an option to enable the feature
+2. Ignore errors which occur both on the `old` and `new` objects
+3. Return ignored errors in a separate list
+
+### `kube-openapi` changes
+
+The `kube-openapi` schema validator validates the standard OpenAPI/JSONSchema
+elements of the CRD schema. This is basically everything except CEL rules,
+and listset/map types.
+
+The current SchemaValidator will need to be updated to add a new sibling to
+its [`Validate(obj) *Result`](https://github.com/kubernetes/kube-openapi/blob/7562a10069617a0c05535f22bf1561b1bce67869/pkg/validation/validate/schema.go#LL109C46-L109C46) method: `ValidateUpdate(old, new) *UpdateResult`
+
+All the supporting validation code within kube-openapi will need to be refactored
+to support both creates and updates, and in the case of an update, to evaluate
+whether an error should be ignored.
+
+`ValidateUpdate`'s `UpdateResult` should return two separate lists: the
+hard validation errors, and any ignored errors.
+
+
+#### Nested Value Validations (anyOf, allOf)
+
+OpenAPI-schemas support nested validations:
+- not (disallowed)
+- oneOf (disallowed)
+- anyOf
+- allOf
+
+
+In the case of `anyOf` and `allOf`, we can support ratcheting these rules. The 
+schema validator will consider the match to be successful even if it has
+ignored errors due to ratcheting.
+
+### Structural-Schema-based validation changes
+
+The two structural-schema based validation methods will need to
+be refactored to support `ValidateUpdate` returning separate lists of errors.
+
+Alternatively, these could be rewritten into kube-openapi, and replaced with
+a validator that a schema is structural.
+
+[`ValidateListSetsAndMaps(fldPath *field.Path, s *schema.Structural, obj map[string]interface{}) field.ErrorList`](https://github.com/kubernetes/kubernetes/blob/694698ca3801fe5aa258d58c1e27ac7f88c29d35/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/listtype/validation.go#L28)
+
+[`objectmeta.Validate(pth *field.Path, obj interface{}, s *structuralschema.Structural, isResourceRoot bool) field.ErrorList`](https://github.com/kubernetes/kubernetes/blob/92042fe6eac61b6dd4092f5c6737050a4a1c43e5/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta/validation.go#L32)
+
+
+#### Normalization
+
+For comparison of `old`/`new` `maps`/`sets`, the objects will need to transform
+into a normal form to make them correlatable. In the case of maps this is simple,
+for the case of sets, the sets will need to be sorted which is possible due to 
+fact that only sets of scalars are supported by Kubernetes.
```

Comment:
if we move the ratcheting logic into a structural schema recursion, we can carry around a path mapping function that is computed for sets and maps beforehand. That would avoid doing an actual normal form copy in memory.

Reply:
xref https://github.com/kubernetes/kubernetes/pull/108073 how old and new schemas are correlated for CEL's "oldSelf" variable.

Reply:
yes, something along these lines.

Reply:
The text still talks about normal forms. I don't think we want to compute normal forms, but use mappings.

Reply:
Changed the text to explicitly say that the implementation will need a performant way to correlate old/new across lists/maps and extended kubernetes types whether that is normal forms or mappings, that is undefined, as long as it works and fits the performance goals.